orca/orca.go

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package orca implements Open Request Cost Aggregation, which is an open
+// standard for request cost aggregation and reporting by backends and the
+// corresponding aggregation of such reports by L7 load balancers (such as
+// Envoy) on the data plane. In a proxyless world with gRPC enabled
+// applications, aggregation of such reports will be done by the gRPC client.
+//
+// Experimental
+//
+// Notice: All APIs is this package are EXPERIMENTAL and may be changed or
+// removed in a later release.
+package orca
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+
+	v3orcapb "github.com/cncf/xds/go/xds/data/orca/v3"
+)
+
+var (
+	logger            = grpclog.Component("orca-backend-metrics")
+	joinServerOptions = internal.JoinServerOptions.(func(...grpc.ServerOption) grpc.ServerOption)
+)
+
+const trailerMetadataKey = "endpoint-load-metrics-bin"
+
+// CallMetricsServerOption returns a server option which enables the reporting
+// of per-RPC custom backend metrics for unary and streaming RPCs.
+//
+// Server applications interested in injecting custom backend metrics should
+// pass the server option returned from this function as the first argument to
+// grpc.NewServer().
+//
+// Subsequently, server RPC handlers can retrieve a reference to the RPC
+// specific custom metrics recorder [MetricSetter] to be used, via a call to
+// MetricSetterFromContext(), and inject custom metrics at any time during the
+// RPC lifecycle.
+//
+// The injected custom metrics will be sent as part of trailer metadata, as a
+// binary-encoded [ORCA LoadReport] protobuf message, with the metadata key
+// being set be "endpoint-load-metrics-bin".
+//
+// [ORCA LoadReport]: https://github.com/cncf/xds/blob/main/xds/data/orca/v3/orca_load_report.proto#L15
+func CallMetricsServerOption() grpc.ServerOption {
+	return joinServerOptions(grpc.ChainUnaryInterceptor(unaryInt), grpc.ChainStreamInterceptor(streamInt))
+}
+
+func unaryInt(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	recorder := newMetricRecorder()
```

Comment:
Is this expensive, in particular the allocations?  Is there something we can do to optimize that doesn't make the code too obfuscated, like lazy initialization or object reuse?  I'm assuming there's no way to allocate it on the stack, but that would be nice, too.

Reply:
Done. Switched to lazy allocation.